// @ceramicnetwork/http-client/lib/remote-admin-api.js
import { fetchJson } from '@ceramicnetwork/common';
import { RemotePinApi } from './remote-pin-api.js';
import { StreamID } from '@ceramicnetwork/streamid';
import { MissingDIDError } from './utils.js';
export class RemoteAdminApi {
    constructor(_apiUrl, _getDidFn) {
        this._apiUrl = _apiUrl;
        this._getDidFn = _getDidFn;
        this._fetchJson = fetchJson;
        this.modelsPath = './admin/models';
        this.getCodePath = './admin/getCode';
        this.nodeStatusPath = './admin/status';
        this._pinApi = new RemotePinApi(this._apiUrl, this._getDidFn);
    }
    getCodeUrl() {
        return new URL(this.getCodePath, this._apiUrl);
    }
    getModelsUrl() {
        return new URL(this.modelsPath, this._apiUrl);
    }
    getStatusUrl() {
        return new URL(this.nodeStatusPath, this._apiUrl);
    }
    async buildJWS(actingDid, code, requestPath, modelsIDs) {
        if (!actingDid)
            throw new MissingDIDError();
        const body = modelsIDs
            ? { models: modelsIDs.map((streamID) => streamID.toString()) }
            : undefined;
        const jws = await actingDid.createJWS({
            code: code,
            requestPath,
            requestBody: body,
        });
        return `${jws.signatures[0].protected}.${jws.payload}.${jws.signatures[0].signature}`;
    }
    async generateCode() {
        return (await this._fetchJson(this.getCodeUrl())).code;
    }
    async nodeStatus() {
        const code = await this.generateCode();
        return this._fetchJson(this.getStatusUrl(), {
            headers: {
                Authorization: `Basic ${await this.buildJWS(this._getDidFn(), code, this.getStatusUrl().pathname)}`,
            },
        });
    }
    async startIndexingModels(modelsIDs) {
        const code = await this.generateCode();
        await this._fetchJson(this.getModelsUrl(), {
            method: 'post',
            body: {
                jws: await this.buildJWS(this._getDidFn(), code, this.getModelsUrl().pathname, modelsIDs),
            },
        });
    }
    async getIndexedModels() {
        const code = await this.generateCode();
        const response = await this._fetchJson(this.getModelsUrl(), {
            headers: {
                Authorization: `Basic ${await this.buildJWS(this._getDidFn(), code, this.getModelsUrl().pathname)}`,
            },
        });
        return response.models.map((modelStreamIDString) => {
            return StreamID.fromString(modelStreamIDString);
        });
    }
    async stopIndexingModels(modelsIDs) {
        const code = await this.generateCode();
        await this._fetchJson(this.getModelsUrl(), {
            method: 'delete',
            body: {
                jws: await this.buildJWS(this._getDidFn(), code, this.getModelsUrl().pathname, modelsIDs),
            },
        });
    }
    get pin() {
        return this._pinApi;
    }
}
